lib/api/core/pluginsManager.js

```diff
@@ -0,0 +1,326 @@
+
+      if (plugin.object.hooks) {
+        _.forEach(plugin.object.hooks, function (fn, event) {
+          if (plugin.object[fn]) {
+            kuzzle.on(event, function (message) {
+              plugin.object[fn](message, event);
+            });
+          }
+        });
+      }
+
+      if (plugin.object.pipes) {
+        _.forEach(plugin.object.pipes, function (fn, pipe) {
+          if (plugin.object[fn]) {
+            if (!this.pipes[pipe]) {
+              this.pipes[pipe] = [];
+            }
+
+            this.pipes[pipe].push(plugin.object[fn]);
+          }
+        }.bind(this));
+      }
+
+    }.bind(this));
+  };
+
+  /**
+   * Trigger an event for emit event and chain pipes
+   *
+   * @param event
+   * @param data
+   * @returns {Promise}
+   */
+  this.trigger = function (event, data) {
+    if (this.isDummy) {
+      this.kuzzle.emit(event, data);
+      return Promise.resolve(data);
+    }
+
+    return triggerPipes.call(this, event, data)
+      .then(function (modifiedData) {
+        return triggerHooks.call(this, event, modifiedData);
+      }.bind(this));
+  };
+};
+
+/**
+ * Emit event
+ * @param event
+ * @param data
+ * @returns {Promise}
+ */
+var triggerHooks = function (event, data) {
+  this.kuzzle.emit(event, data);
+
+  return Promise.resolve(data);
+};
+
+/**
+ * Chain call all attached functions plugins on the specific event
+ * @param event
+ * @param data
+ * @returns {Promise}
+ */
+var triggerPipes = function (event, data) {
+  var
+    deferred = q.defer(),
+    preparedPipes = [],
+    indexDelimiter,
+    wildcard;
+
+  indexDelimiter = event.indexOf(':');
+  if (indexDelimiter !== 1) {
+    wildcard = event.substring(0, indexDelimiter+1) + '*';
+  }
+
+  if (this.pipes && this.pipes[event] && this.pipes[event].length) {
+    preparedPipes = this.pipes[event];
+  }
+
+  if (wildcard && this.pipes && this.pipes[wildcard] && this.pipes[wildcard].length) {
+    preparedPipes = preparedPipes.concat(this.pipes[wildcard]);
+  }
+
+  if (preparedPipes.length > 0) {
+    // Add a fake function in first position for return the 'data' to other chained functions
+    preparedPipes.unshift(function (callback) { callback(null, data); });
+  }
+  else {
+    deferred.resolve(data);
+    return deferred.promise;
+  }
+
+  async.waterfall(preparedPipes, function (error, result) {
+    deferred.resolve(result);
+  });
+
+  return deferred.promise;
+};
+
+/**
+ * Install given plugins
+ * @param plugins
+ * @param basePlugins
+ * @returns {boolean}
+ */
+var installPlugins = function (plugins, basePlugins) {
+  var
+    newInstalled = false,
+    pluginInstallId;
+
+  _.forEach(plugins, function (plugin, name) {
+    if (plugin.url) {
+      pluginInstallId = plugin.url;
+    }
+    else if (plugin.version) {
+      pluginInstallId = name + '@' + plugin.version;
+    }
+    else {
+      this.kuzzle.log.error('Plugin', name, 'has no version. The version is mandatory if there is no URL.');
+      return true;
```

Comment:
We return true because we're in a loop and we want to pass to the next iteration. In lodash, if we return false in `forEach`, the loop is stoped.
The installation status is `newInstalled` and it's return after the loop.